File: Java/Linked List Cycle.java
public class Solution {

    /**
     * Time:   O(n)
     * Memory: O(1)
     */
    public boolean hasCycle(ListNode head) {
        ListNode slow = head, fast= head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next;

            if (slow == fast)
                return true;
        }

        return false;
    }
}

File: Python/Linked List Cycle.py
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    """
    Time:   O(n)
    Memory: O(1)
    """

    def hasCycle(self, head: Optional[ListNode]) -> bool:
        fast = head
        slow = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True

        return False
